src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,498 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   Bitmapset *conkeys, Oid conrelid)
+{
+	BOOL result = false;
+	Bitmapset *conkeys_copy = gpdb::BmsCopy(conkeys);
+	ListCell *lgc;
+
+	// Search for conkey in grouping list.
+	// Once a conkey is found, remove it from a copy of Bitmapset conkeys.
+	// If conkeys_copy set is empty at the end, then grouping_list contains
+	// full set of columns, that define the primary key.
+	ForEach(lgc, grouping_list)
+	{
+		Node *grpcl = (Node *) lfirst(lgc);
+
+		if (NULL == grpcl || !IsA(grpcl, SortGroupClause))
+		{
+			continue;
+		}
+
+		SortGroupClause *sgc = (SortGroupClause *) grpcl;
+		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+		GPOS_ASSERT(IsA(expr, Var));
+
+		Var *var = (Var *) expr;
+
+		RangeTblEntry *rte =
+			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+		GPOS_ASSERT(NULL != rte);
+
+		if (gpdb::BmsIsMember(var->varattno, conkeys_copy) &&
+			var->varlevelsup == 0 && rte->relid == conrelid)
+		{
+			conkeys_copy = gpdb::BmsDelMember(conkeys_copy, var->varattno);
```

Comment:
I think some kind of distinction / difference between the two bitmaps would be better suited here.